rfcs/v3/authorization-rules.md

```diff
@@ -0,0 +1,568 @@
+- SupportTickets: id, assigned_agent_id
+
+And these are the potential authorization considerations:
+
+- Two personas: Developers and Support Agents
+  - Developers can generally view all user's information
+  - Support agents can only view data for users whose tickets they have been assigned
+- Some user information is protected
+  - PII fields (name, email) unless the API consumer has explicit access
+  - Government user data is restricted unless the API consumer has explicit access
+
+### Before
+
+You would need to define the following roles and provision them appropriately in the authentication system.
+
+- developer_base
+- developer_with_pii_access
+- developer_with_gov_access
+- developer_with_pii_access_and_gov_access
+- support_agent_base
+- support_agent_with_pii_access
+- support_agent_with_gov_access
+- support_agent_with_pii_access_and_gov_access
+
+And the permissions would be very hard to manage. Eg:
+
+```yaml
+kind: TypePermission
+version: v1
+definition:
+  typeName: User
+  roles:
+    - role: developer_base
+      output:
+        allowedFields: [id]
+    - role: developer_with_pii_access
+      output:
+        allowedFields: [id, name, email]
+    - role: developer_with_gov_access
+      output:
+        allowedFields: [id]
+    - role: developer_with_pii_access_and_gov_access
+      output:
+        allowedFields: [id, name, email]
+    - role: support_agent_base
+      output:
+        allowedFields: [id]
+    - role: support_agent_with_pii_access
+      output:
+        allowedFields: [id, name, email]
+    - role: support_agent_with_gov_access
+      output:
+        allowedFields: [id]
+    - role: support_agent_with_pii_access_and_gov_access
+      output:
+        allowedFields: [id, name, email]
+```
+
+The model permissions for User and UserActivity would be massive. For example:
+
+```yaml
+kind: ModelPermissions
+version: v1
+definition:
+  modelName: User
+  roles:
+    - role: developer_base # Same for developer_with_pii_access
+      select:
+        filter:
+          not:
+            fieldComparison:
+              fieldName: is_gov
+              operator: _eq
+              value: literal: true
+    - role: developer_with_gov_access # Same for developer_with_pii_access_and_gov_access
+      select:
+        filter: null
+    - role: support_agent_base # Same for support_agent_with_pii_access
+      select:
+        filter:
+          and:
+            - relationship:
+                name: tickets
+                predicate:
+                  fieldComparison:
+                    field: agent_id
+                    operator: _eq
+                    value: sessionVariable: x-hasura-agent-id
+            - not:
+                fieldComparison:
+                  fieldName: is_gov
+                  operator: _eq
+                  value: literal: true
+      - role: support_agent_with_gov_access # Same for support_agent_with_pii_access_and_gov_access
```

Comment:
If we're concerned about the usability of metadata, then it doesn't seem wild that eventually we're going to want to abstract these predicates too and refer to them by name.

```yaml
.....
- role: support_agent_with_gov_access # Same for support_agent_with_pii_access_and_gov_access
  select:
    filter:
      relationship:
        name: tickets
        predicate: agent_predicate
---
kind: Predicate
version: v1
definition:
  name: agent_predicate
  fieldComparison:
    field: agent_id
    operator: _eq
    value: sessionVariable: x-hasura-agent-id
```

Reply:
In a world where my rules are composable, the ability to reuse predicates isn't that important.
The fundamental problem that is being addressed is not the ability to reuse parts of my metadata, but not having to enumerate all possible authorization states in my system.